odl/space/base_tensors.py

```diff
@@ -0,0 +1,741 @@
+﻿# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Base classes for implementations of tensor spaces."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+import numpy as np
+
+from odl.set.sets import Set, RealNumbers, ComplexNumbers
+from odl.set.space import LinearSpace, LinearSpaceElement
+from odl.util import (
+    is_scalar_dtype, is_floating_dtype, is_real_dtype,
+    is_complex_floating_dtype, safe_int_conv,
+    arraynd_repr, arraynd_str, dtype_str, signature_string, indent_rows)
+from odl.util.ufuncs import TensorSpaceUfuncs
+from odl.util.utility import TYPE_MAP_R2C, TYPE_MAP_C2R
+
+
+__all__ = ('TensorSpace',)
```

Comment:
I think the fact that we don't want users to initiate such an object directly is a strong reason for not exposing it at top level.